### ISSUE TITLE
11.7 Benutzerpräferenzen bei webview

### ISSUE THREAD

**Issue by stefanFarnetani** [BIK-BITV/BIK-App-Test#131]
So viel ich weiß, sind Schriftgrößenanpassungen aufgrund der Einstellungen in den Settings der Smartphones in "Webviews" technisch nicht möglich.

Unter iOS gibt es eine recht umständliche Art es hinzubekommen. Aber ich glaube, dieses Vorgehen schränkt Umsetzer zu stark ein und wird deshalb nie gemacht. Man muss neben der Größe auch die Schriftart von iOS übernehmen.

Mein Vorschlag wäre unter "2. Prüfung" einen Abschnitt aufzunehmen:
===== Webview

Bei einem Webviews mit ausgeblendeten WEb-UI (URL-Zeile und Browser-Funktionen) wird der Prüfschritt als "nicht anwendbar"  bewertet.
Hintergrund ist die fehlende technische Möglichkeit in Webviews auf Einstellungen im Betriebsystem zu reagieren.

**Comment by anatom5** [comment]
Wenn ich Deinen Kommentar richtig verstehe, bezieht sich das auf native Apps, oder? Also eine native iOS- oder Android-App, die Web-Inhalte in einen eingebetteten Browser anzeigt. Das ist ja nicht Teil von BIK BITV- / WCAG-Test (Web). Oder verstehe ich Dich gerade falsch?

**Comment by stefanFarnetani** [comment]
@anatom5 ich kommentiere hier im repository BIK App Test. Der soll doch mobile Apps testen. oder?

**Comment by sweckenmann** [comment]
@stefanFarnetani Ja, du bist richtig beim App-Test. Und ja, der testet mobile Apps.

**Comment by stefanFarnetani** [comment]
@sweckenmann danke.
@anatom5 Ja dann stimmt das alles. Es gibt vereinfacht gesagt drei arten eine (native) mobile App zu programmieren.
1. nativ programmieren je Plattform
2. ein Framework verwenden (z.B. react native), dass den code jeweils in nativen Code umwandelt
3. in HTML/JS/CSS programmieren und nur einen nativen Container verwenden

Beim letzteren hat man im Endeffekt eine HTML-Seite. Mit dem Malus, dass man keinen Browser-GUI mehr hat. Damit hat man alle Nachteile der HTML-Seite (kein direkter zugriff auf eingestellte Schriftgrößen) und den Nachteil, dass Nutzer keine Funktion zum manuellen Vergrößern mehr haben.

Es empfinde es, als seltsam jemanden zu bestrafen, für eine gewählte technische Umsetzung. Es gibt auch Unterschiede zwischen den Funktionen von iOS und Android und wir akzeptieren diese in anderen Prüfschritten ohne Abzug.

Aber ich gebe auch zu, dass es schwer ist herauszufinden, ob eine HTML-Seite in einem Container steckt. Das erkennt man nur der Vergabe von Rollen und Attribute, die es auf mobilen Geräten nicht gibt.

**Comment by detlevhfischer** [comment]
> Aber ich gebe auch zu, dass es schwer ist herauszufinden, ob eine HTML-Seite in einem Container steckt. Das erkennt man nur der Vergabe von Rollen und Attribute, die es auf mobilen Geräten nicht gibt.

@stefanFarnetani Wenn es auf der Webseite Auszeichnungen gibt werden diese i.d.R. von der App ausgegeben, einschließlich Überschriftenhierarchien und ARIA-Auszeichnungen.

**Comment by sprungmarker** [comment]
@stefanFarnetani  ja, ich habe auch aktuell einen Kunden, der genau das Problem hat. Auf der Webversion greift der Pagezoom. Die gleiche Ansicht wird als Webview in der nativen App integriet und dort wird der Textzoom getestet. Was natürlich zu groben Problemen führt.

Es gäbe ja auch die Möglichkeit, in einer nativen App auf einen reinen App-Zoom zurückzugreifen, ählich wie im Web und nicht explizit nur den Text zu vergrößern.

Danke, @detlevhfischer ich schaue mir die Debatte auf Etsi an - das Problem ich muss aktuell das beraten und die Kunden haben halt auch das Gefühl, dass Webviews als Option irgendwie benachteiligt werden im App-Kontext.

**Comment by sprungmarker** [comment]
Der aktuelle Vorschlag der Konkretisierung scheint Webviews für diesen Prüfschritt doch auszuschliessen oder?

https://labs.etsi.org/rep/HF/en301549/-/issues/23#note_18028

```
11.7 User preferences

Where ICT is or includes non-web software that provides a user interface, and is not designed to be isolated from its platform,
that user interface shall follow the values of the preferences that users have set for the platform documented accessibility features.

NOTE 1:	Software that is isolated from its underlying platform has no access to user settings in the platform and thus cannot adhere to them.
```

**Comment by stefanFarnetani** [comment]
Hi @sprungmarker 

So interpretiere ich die überarbeiteten Notes nicht.
"isolated from its underlying platform" wäre für mich sowas wie ein geschlossenes Kiosk-System oder eine Spielkonosle. Es wird vom Nutzer aus betrachtet. Dieser hat keinen Zugriff auf das darunter liegende Betriebssystem. In einer App auf meinem Smartphone habe ich als Nutzer Zugriff auf Einstellungen des Betriebsystems.

Auch wenn durch die Notes einiges klarer definiert wurde, ist der Prüfschritt 11.7 immer noch sehr vage.

Mit "non-web software" ist bereits an anderer Stelle geklärt worden (finde die Diskussion gerade nicht mehr), dass alles ohne sichtbare Browseer-GUI und URL-Zeile als non-web software zählt.

In meinen Augen sind wir in Bezug auf unsere Fragestellung, keinen Schritt weiter nach der Überarbeitung von 11.7.
Sagt mir, wenn ich etwas übersehe. Ich finde die aktuelle Situation unzumutbar für die Softwareentwicklung. Webviews in mobilen Apps sind sehr wichtig. Es gibt viele Nachvollziehbare use-cases. 
Und es existiert einfach keine Techniken webviews nach 11.7 barrierefrei zu bekommen.

**Comment by sprungmarker** [comment]
@stefanFarnetani ja, wie ich ja schon ausführte, müssen Kunden, wenn sie Webviews einsetzen - ein doppelte Herangehensweise ans Zooming machen - als auch den Textzoom an sich komplett umsetzen, damit es in der Verwendung in einer App passt. Das finde ich schon zu viel verlangt.

**Comment by stefanFarnetani** [comment]
> [@stefanFarnetani](https://github.com/stefanFarnetani) ja, wie ich ja schon ausführte, müssen Kunden, wenn sie Webviews einsetzen - ein doppelte Herangehensweise ans Zooming machen - als auch den Textzoom an sich komplett umsetzen, damit es in der Verwendung in einer App passt. Das finde ich schon zu viel verlangt.

@sprungmarker ich bin ganz bei dir. Es ist viel verlangt und ich schlage weiterhin vor, dass wir es als "nicht anwendbar" betrachten. Aber tatsächlich aus einer technischen Lücke heraus. Das mache ich nicht gerne. Aber ich sehe es aktuell nicht anders möglich.
Der überarbeitete EN-Text hat in dieser Sache keine Verbesserung gebracht. Ich habe die Diskussion überflogen. Webviews waren leider kein Thema. Weiß jemand, ob man noch Vorschläge einreichen kann und Wo? Der Zeitplan war ja bereits recht weit fortgeschritten für die Überarbeitung der EN.

**Comment by detlevhfischer** [comment]
@stefanFarnetani @sprungmarker Hier ein von input von @JJdeGroot auf eine Frage von mir zur Umsetzbarkeit von 1.4.4 in Webviews:
https://github.com/w3c/matf/issues/3#issuecomment-2637489160

**Comment by stefanFarnetani** [comment]
@detlevhfischer Danke, dass du die Frage ins größere Forum getragen hast.

Entschuldigt die späte Antwort. Viel zu tun ... wie wahrscheinlich bei allen aktuell.

Ich möchte ganz kurz auf die von Jan Jaap de Groot genannten Möglichkeiten besprechen:

> Some options:
> 1. Pass the text scaling multiplier using a custom HTTP header. Only works if you are showing 1st party content because you need to apply CSS/JavaScript based on this header.
> 2. Apply text scaling using custom css (iOS: font: -apple-system-body; not sure if Android has equivalent). This can work for any content, also 3rd party - but might behave unexpectedly.
> 3. Implement a zoom button in your host app, that applies zoom to the WebView (like a browser app)

Zu Lösung 3. Haben wir auch besprochen. Ist halt sehr viel verlangt und löst nur Text-zoom aber nicht 11.7

Zu Lösung  2. Ist in meinen Augen keine angemessene Lösung. Funktioniert nur auf iOS. Außerdem muss man sich auch auf die Standard-Schriftart das System einlassen ansonsten funktioniert es nicht. Und löst ebenfalls nicht 11.7.

Zu Lösung 1. In meinen Augen bleibt damit nur eine Lösung. Auf diese Idee mit dem HTTP Header bin ich tatsächlich bisher nicht gekommen. Es fühlt sich auch "hacky" an. Für alle nicht Techniker, man missbraucht sozusagen den HTTP-Header, um da eigene Informationen zu transportieren. Diese kann man dann beim Server wieder abrufen. Setzt voraus, dass man den Quellcode der eingebetteten Seite kontrolliert. Die Kontrolle der Webview-Seite kann man aber an dieser Stelle voraussetzten. Handelt es sich um eine Fremdseite, soll besser die Seite im Browser des Smartphones geöffnet werden. 

Ich würde die Lösung als Hack bezeichnen. Aber es wird funktionieren. Ich habe es selbst nicht ausprobiert, aber ich sehe keinen Grund wieso es nicht prinzipiell funktionieren sollte. 
Mir schmeckt es nicht, dass wir auf so einen Hack ausweichen müssen, um das zu ermöglichen. Aber es gibt damit eine Möglichkeit wie man das erreichen kann. 


Zusammenfassend:
Ich würde damit Vorschlagen, dass die Frage beantwortet ist.
11.7 ist auch bei eingebetteten Webviews in mobilen App anwendbar, da es zumindest eine Lösung gibt, mit der es machbar ist.
